packages/flutter/lib/src/rendering/slivers.dart

```diff
@@ -0,0 +1,1850 @@
+  }
+
+  @override
+  ScrollDirection get userScrollDirection => ScrollDirection.idle;
+}
+
+// ///
+// /// See also:
+// ///
+// /// - [RenderSliver], which explains more about the Sliver protocol.
+// /// - [RenderBox], which explains more about the Box protocol.
+// /// - [RenderSliverToBoxAdapter], which allows a [RenderBox] object to be
+// ///   placed inside a [RenderSliver] (the opposite of this class).
+class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSliver, SliverPhysicalContainerParentData> {
+  /// Creates a viewport for [RenderSliver] objects.
+  ///
+  /// If the [center] is not specified, then the first child in the `children`
+  /// list, if any, is used.
+  RenderViewport2({
+    AxisDirection axisDirection: AxisDirection.down,
+    bool shrinkWrap: false,
+    double anchor: 0.0,
+    ViewportOffset offset,
+    List<RenderSliver> children,
+    RenderSliver center,
+  }) : _axisDirection = axisDirection,
+       _shrinkWrap = shrinkWrap,
+       _anchor = anchor,
+       _offset = offset ?? new ViewportOffset.zero(),
+       _center = center {
+    addAll(children);
+    if (center == null && firstChild != null)
+      _center = firstChild;
+    assert(axisDirection != null);
+    assert(shrinkWrap != null);
+    assert(anchor != null);
+    assert(anchor >= 0.0 && anchor <= 1.0);
+  }
+
+  AxisDirection get axisDirection => _axisDirection;
+  AxisDirection _axisDirection;
+  set axisDirection(AxisDirection value) {
+    assert(value != null);
+    if (value == _axisDirection)
+      return;
+    _axisDirection = value;
+    markNeedsLayout();
+  }
+
+  Axis get axis {
+    assert(axisDirection != null);
+    switch (axisDirection) {
+      case AxisDirection.up:
+      case AxisDirection.down:
+        return Axis.vertical;
+      case AxisDirection.left:
+      case AxisDirection.right:
+        return Axis.horizontal;
+    }
+    return null;
+  }
+
+  bool get shrinkWrap => _shrinkWrap;
+  bool _shrinkWrap;
+  set shrinkWrap(bool value) {
+    assert(value != null);
+    if (value == _shrinkWrap)
+      return;
+    _shrinkWrap = value;
+    markNeedsLayout();
+  }
+
+  double get anchor => _anchor;
+  double _anchor;
+  set anchor(double value) {
+    assert(value != null);
+    assert(value >= 0.0 && value <= 1.0);
+    if (value == _anchor)
+      return;
+    _anchor = value;
+    markNeedsLayout();
+  }
+
+  ViewportOffset get offset => _offset;
+  ViewportOffset _offset;
+  set offset(ViewportOffset value) {
+    value ??= new ViewportOffset.zero();
+    if (value == _offset)
+      return;
+    if (attached)
+      _offset.removeListener(markNeedsLayout);
+    if (_offset.pixels != value.pixels)
+      markNeedsLayout();
+    _offset = value;
+    if (attached)
+      _offset.addListener(markNeedsLayout);
+  }
+
+  RenderSliver get center => _center;
+  RenderSliver _center;
+  set center(RenderSliver value) {
+    if (value == _center)
+      return;
+    _center = value;
```

Comment:
It's validated in `performLayout`. I don't want to validate it here because that would prevent you from setting the center pointer before adding the child, which, in addition to being annoying in general, would be particularly annoying for us at the Widgets layer IIRC.